### PR TITLE
Add energy correlation functions to jet tools

### PR DESCRIPTION
**Pull request by ahinzmann**
The energy correlations functions are currently the default algorithm for W-tagging in ATLAS.
They are not run in RECO/AOD/MiniAOD, but added them to CMSSW for cross check analyses.
They can be tested with:
RecoJets/JetProducers/test/testJetTools_cfg.py

### CODE REVIEW

**Comment by cvuosalo** [comment]
@cmsbuild please test

File: RecoJets/JetProducers/interface/ECFAdder.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#ifndef ECFAdder_h
```
Comment (slava77) [inline RecoJets/JetProducers/interface/ECFAdder.h]: add package name to the include guard

File: RecoJets/JetProducers/interface/ECFAdder.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/JetReco/interface/Jet.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "fastjet/contrib/EnergyCorrelator.hh"
+
+
+class ECFAdder : public edm::EDProducer {
```
Comment (slava77) [inline RecoJets/JetProducers/interface/ECFAdder.h]: this should be made multithreading-friendly
choice will depending on thread-safety of the EnergyCorrelator.
@Dr15Jones is it on our bad or good list?

File: RecoJets/JetProducers/interface/ECFAdder.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include "fastjet/contrib/EnergyCorrelator.hh"
+
+
+class ECFAdder : public edm::EDProducer { 
+ public:
+    explicit ECFAdder(const edm::ParameterSet& iConfig);
+    
+    virtual ~ECFAdder() {}
+    
+    void produce(edm::Event & iEvent, const edm::EventSetup & iSetup) ;
```
Comment (slava77) [inline RecoJets/JetProducers/interface/ECFAdder.h]: add ``` override ``` at the end

File: RecoJets/JetProducers/interface/ECFAdder.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#include "DataFormats/JetReco/interface/Jet.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "fastjet/contrib/EnergyCorrelator.hh"
+
+
+class ECFAdder : public edm::EDProducer { 
+ public:
+    explicit ECFAdder(const edm::ParameterSet& iConfig);
+    
+    virtual ~ECFAdder() {}
```
Comment (slava77) [inline RecoJets/JetProducers/interface/ECFAdder.h]: empty destructors are not needed in general

File: RecoJets/JetProducers/plugins/ECFAdder.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ECFAdder::ECFAdder(const edm::ParameterSet& iConfig) :
+  src_(iConfig.getParameter<edm::InputTag>("src")),
+  src_token_(consumes<edm::View<reco::Jet>>(src_)),
+  Njets_(iConfig.getParameter<std::vector<unsigned> >("Njets")),
+  beta_(iConfig.getParameter<double>("beta"))
+{
+  for ( std::vector<unsigned>::const_iterator n = Njets_.begin(); n != Njets_.end(); ++n )
+    {
+      std::ostringstream ecfN_str;
+      ecfN_str << "ecf" << *n;
```
Comment (slava77) [inline RecoJets/JetProducers/plugins/ECFAdder.cc]: why not just ``` string a = "ecf" + *n ``` ?

File: RecoJets/JetProducers/plugins/ECFAdder.cc [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+void ECFAdder::produce(edm::Event & iEvent, const edm::EventSetup & iSetup) {
+  // read input collection
+  edm::Handle<edm::View<reco::Jet> > jets;
+  iEvent.getByToken(src_token_, jets);
+  
+  for ( std::vector<unsigned>::const_iterator n = Njets_.begin(); n != Njets_.end(); ++n )
+    {
+      std::ostringstream ecfN_str;
```
Comment (slava77) [inline RecoJets/JetProducers/plugins/ECFAdder.cc]: don't recompute these every event. These can be filled after config parsing.

File: RecoJets/JetProducers/python/ecfAdder_cfi.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import FWCore.ParameterSet.Config as cms
+
+ECF = cms.EDProducer("ECFAdder",
```
Comment (slava77) [inline RecoJets/JetProducers/python/ecfAdder_cfi.py]: some longer and more intuitive name would be better.
Our rules are to name the module instance and the cfi file with the same name.
I like ecfAdder more than ECF (even though ecf could've been more descriptive).

File: RecoJets/JetProducers/interface/ECFAdder.h (hunk 2 above)
Comment (Dr15Jones) [inline RecoJets/JetProducers/interface/ECFAdder.h]: I took a quick look at that code and as far as I can tell it is thread friendly and can be used with a stream module.

File: RecoJets/JetProducers/python/ecfAdder_cfi.py (hunk 7 above)
Comment (ahinzmann) [inline RecoJets/JetProducers/python/ecfAdder_cfi.py]: I go with ECF to be synchronized with Njettiness. In this way the (potenial) userdata name in MiniAOD is simpler.

File: RecoJets/JetProducers/plugins/ECFAdder.cc (hunk 5 above)
Comment (ahinzmann) [inline RecoJets/JetProducers/plugins/ECFAdder.cc]: this didn't work for me:
std::string ecfN_str="ecf" + *n;
produces<edm::ValueMap<float> >(ecfN_str.c_str());
so I didn't change it.
This is however only done in the constructor.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #11028 b98303e20e85af4da22a27de9d7119d4c74157b2

Adding energy correlation functions to jet tools. This new tool doesn't run in standard workflows, so there should be no changes in run-time behavior.

The new code is satisfactory. Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11028/7766/summary.html) against baseline CMSSW_7_6_X_2015-08-30-2300 show no significant differences, except for DQM differences caused by #10923, which is included in the IB. These latter differences are not related to this PR. A check of the new jet test script shows it runs successfully and produces a Root file.

**Comment by davidlange6** [comment]
+1

**Pull request merged**